### ISSUE TITLE
bug report: function latent_optimise in losses.py

### ISSUE THREAD

**Issue by lihuiknight** [POSTECH-CVLab/PyTorch-StudioGAN#63]
the `transport_cost`  is not defined before it has been used in [line ](https://github.com/POSTECH-CVLab/PyTorch-StudioGAN/blob/master/src/utils/losses.py#L71)

**Comment by MINGUKKANG** [comment]
Hi, 

Can you give me a training config that raises the issue above?

Thank you.

**Comment by lihuiknight** [comment]
I didn't run this function. I am reading your implementation, and find this potential bug. Have a check you can.

**Comment by MINGUKKANG** [comment]
Hello,

The transport_cost variable is defined when step is 0.

Please refer to the below code.

```python3
        if trans_cost:
            if step == 0:
                transport_cost = (delta_z.norm(2, dim=1)**2).mean()
            else:
                transport_cost += (delta_z.norm(2, dim=1)**2).mean()
            return zs, trans_cost
        else:
            return zs
```

I really appreciate your bug reporting.

Thank you very much.

Best,

Minguk Kang

**Comment by lihuiknight** [comment]
yes, you are right. thanks for your reply.

**Comment by MINGUKKANG** [comment]
Hi Ken Li,

Thank to your report, I have found a critical bug in the latent_optimise function.

I think this is why StudioGAN gives poor generation results from LOGAN experiments.

Thank you so much.

Best,

Minguk 



Before:
```python
def latent_optimise(zs, fake_labels, gen_model, dis_model, conditional_strategy, latent_op_step, latent_op_rate,
                    latent_op_alpha, latent_op_beta, trans_cost, default_device):
    batch_size = zs.shape[0]
    for step in range(latent_op_step):
        drop_mask = (torch.FloatTensor(batch_size, 1).uniform_() > 1 - latent_op_rate).to(default_device)
        z_gradients, z_gradients_norm = calc_derv(zs, fake_labels, dis_model, conditional_strategy, default_device, gen_model)
        delta_z = latent_op_alpha*z_gradients/(latent_op_beta + z_gradients_norm)
        zs = torch.clamp(zs + drop_mask*delta_z, -1.0, 1.0)

        if trans_cost:
            if step == 0:
                transport_cost = (delta_z.norm(2, dim=1)**2).mean()
            else:
                transport_cost += (delta_z.norm(2, dim=1)**2).mean()
            return zs, trans_cost
        else:
            return zs
```

After:
```python
def latent_optimise(zs, fake_labels, gen_model, dis_model, conditional_strategy, latent_op_step, latent_op_rate,
                    latent_op_alpha, latent_op_beta, trans_cost, default_device):
    batch_size = zs.shape[0]
    for step in range(latent_op_step):
        drop_mask = (torch.FloatTensor(batch_size, 1).uniform_() > 1 - latent_op_rate).to(default_device)
        z_gradients, z_gradients_norm = calc_derv(zs, fake_labels, dis_model, conditional_strategy, default_device, gen_model)
        delta_z = latent_op_alpha*z_gradients/(latent_op_beta + z_gradients_norm)
        zs = torch.clamp(zs + drop_mask*delta_z, -1.0, 1.0)

        if trans_cost:
            if step == 0:
                transport_cost = (delta_z.norm(2, dim=1)**2).mean()
            else:
                transport_cost += (delta_z.norm(2, dim=1)**2).mean()

    if trans_cost:
        return zs, trans_cost
    else:
        return zs
```

**Comment by lihuiknight** [comment]
Great.  will you report the performance of LOGAN on CIFAR10, tiny imagenet later?

**Comment by MINGUKKANG** [comment]
Sure:)

**Comment by Nayan-chereddycmd** [comment]
Stopppp

**Comment by MINGUKKANG** [comment]
Hi,

I have corrected the wrong implementation of latent optimization and will report performances of LOGAN ASAP.

Best,

Minguk

**Comment by lihuiknight** [comment]
@MINGUKKANG  I have tested the implementation. There is a bug in [line](https://github.com/POSTECH-CVLab/PyTorch-StudioGAN/blob/45c27a2be30ef135b68225ea77635538b940e1d8/src/utils/sample.py#L31)

You can fix this bug by change 
`latents = truncated_normal([batch_size, dim], truncated_factor).to(device)` 
to 
`latents = torch.FloatTensor(truncated_normal([batch_size, dim], truncated_factor)).to(device)`

Have a check.

**Comment by MINGUKKANG** [comment]
Sorry for late. 

I will check the bug and fix it. 

Thank you.

Best,

Minguk

**Comment by MINGUKKANG** [comment]
Thank you!

Good day:)
